### PR TITLE
Add generate container CI URL for Web App

### PR DESCRIPTION
**Pull request by aspaan**
---

This checklist is used to make sure that common guidelines for a pull request are followed.

### General Guidelines

- [x] The PR has modified HISTORY.rst with an appropriate description of the change (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

### Command Guidelines

- [x] Each command and parameter has a meaningful description.
- [x] Each new command has a test.

(see [Authoring Command Modules](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules))

### CODE REVIEW

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 1]
```diff
@@ -840,6 +840,24 @@ def list_publish_profiles(resource_group_name, name, slot=None):
...
+                                      'list_publishing_profile_xml_with_secrets', slot)
+    full_xml = ''
+    for f in content:
+        full_xml += f.decode()
+
+    profiles = xmltodict.parse(full_xml, xml_attribs=True)['publishData']['publishProfile']
+    url = ''
+    for profile in profiles:
+        if profile['@publishMethod'] == 'MSDeploy':
+            url = 'https://' + profile['@userName'] + ':' + profile['@userPWD'] + '@' + profile['@publishUrl'] + '/docker/hook'  # pylint: disable=line-too-long
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Should we break out the loop here?

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 2]
```diff
@@ -840,6 +840,24 @@ def list_publish_profiles(resource_group_name, name, slot=None):
...
+def list_container_cd_url(resource_group_name, name, slot=None):
+    import xmltodict
+
+    content = _generic_site_operation(resource_group_name, name,
+                                      'list_publishing_profile_xml_with_secrets', slot)
+    full_xml = ''
+    for f in content:
+        full_xml += f.decode()
+
+    profiles = xmltodict.parse(full_xml, xml_attribs=True)['publishData']['publishProfile']
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: you should reuse `list_publish_profiles`

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py [hunk 3]
```diff
@@ -189,6 +189,11 @@
     short-summary: get publishing endpoints, credentials, database connection strings, etc
 """
 
+helps['webapp deployment list-container-cd-url'] = """
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py]: this should be `show...` unless you output an array

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py [hunk 4]
```diff
@@ -189,6 +189,11 @@
...
+    type: command
+    short-summary: get webhook url for your web app on Linux to configure your container repo for continuous deployment setup
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py]: Suggest: ` get the webhook url of the linux webapp you can use to configure the container repository for continuous deployment`

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py [hunk 5]
```diff
@@ -326,6 +326,10 @@ def test_linux_webapp(self, resource_group):
         self.cmd('webapp config set -g {} -n {} --startup-file {}'.format(resource_group, webapp, 'process.json'), checks=[
             JMESPathCheckV2('appCommandLine', 'process.json')
         ])
+
+        result = self.cmd('webapp deployment list-container-cd-url -g {} -n {}'.format(resource_group, webapp)).get_output_in_json()
+        self.assertTrue(result.startswith('https://'))
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py]: Suggest test a bit more details. All urls in CLI/SDK start with 'https://'

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py (hunk 3 above)
Comment (aspaan) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py]: changed

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py (hunk 4 above)
Comment (aspaan) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py]: changed

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py (hunk 2 above)
Comment (aspaan) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: changed

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py (hunk 5 above)
Comment (aspaan) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/tests/test_webapp_commands.py]: added more testing for string end

**Comment by aspaan** [comment]
Hi Yugang, just pinging to see if anything was left here.

File: src/command_modules/azure-cli-appservice/HISTORY.rst [hunk 6]
```diff
@@ -4,6 +4,10 @@ Release History
 ===============
 (unreleased)
 +++++++++++++++++++
+* webapp: Add generate container CI URL for Web App
+
+(unreleased)
```
Comment (derekbekoe) [inline src/command_modules/azure-cli-appservice/HISTORY.rst]: 'unreleased' appears twice.
Please combine the new addition to the already existing header.

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py [hunk 7]
```diff
@@ -189,6 +189,11 @@
     short-summary: get publishing endpoints, credentials, database connection strings, etc
 """
 
+helps['webapp deployment show-container-cd-url'] = """
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/_help.py]: Based on the [discussion](https://github.com/Azure/azure-cli/issues/4023#issuecomment-316490764), the command should be under `webapp deployment container`. Please correct it. Your change in the commands file is right

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/commands.py [hunk 8]
```diff
@@ -126,6 +126,10 @@ def _polish_bad_errors(ex):
 cli_command(__name__, 'webapp deployment user set', custom_path + 'set_deployment_user')
 cli_command(__name__, 'webapp deployment list-publishing-profiles',
             custom_path + 'list_publish_profiles')
+cli_command(__name__, 'webapp deployment container config',
+            custom_path + 'enable_cd')
+cli_command(__name__, 'webapp deployment container show-container-cd-url',
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/commands.py]: Suggest `show-cd-url`. `container` is redundant under a category with the same name

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 9]
```diff
@@ -840,6 +840,35 @@ def list_publish_profiles(resource_group_name, name, slot=None):
     return converted
 
 
+def enable_cd(resource_group_name, name, enable, slot=None):
+    settings = []
+    settings.append("DOCKER_ENABLE_CI=" + enable)
+    update_app_settings(resource_group_name, name, settings, slot)
+    if enable == 'true':
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Suggest remove the condition, so the command can be consistent on output. The format I would suggest
```json
{
   "DOCKER_ENABLE_CI": "true or false"
   "CD_URL" : "<url or empty>" 
}
```

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 10]
```diff
@@ -840,6 +840,35 @@ def list_publish_profiles(resource_group_name, name, slot=None):
...
+
+def show_container_cd_url(resource_group_name, name, slot=None):
+    settings = get_app_settings(resource_group_name, name, slot)
+    docker_enabled = False
+    for setting in settings:
+        if setting['name'] == 'DOCKER_ENABLE_CI' and setting['value'] == 'true':
+            docker_enabled = True
+            break
+
+    if not docker_enabled:
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: I would suggest not throw, returning an empty string is fine. Lots of CLI show commands were updated to not throw on similar situations, so let us be consistent

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 11]
```diff
@@ -840,6 +840,35 @@ def list_publish_profiles(resource_group_name, name, slot=None):
...
+        raise CLIError('continuous integration is not enabled on docker')
+
+    profiles = list_publish_profiles(resource_group_name, name, slot)
+    for profile in profiles:
+        if profile['publishMethod'] == 'MSDeploy':
+            scmUrl = profile['publishUrl'].replace(":443", "")
+            url = 'https://' + profile['userName'] + ':' + profile['userPWD'] + '@' + scmUrl + '/docker/hook'
+            break
+
+    return url
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: Like i commented, I suggested returning an object with 2 fields

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 12]
```diff
@@ -840,6 +840,44 @@ def list_publish_profiles(resource_group_name, name, slot=None):
...
+            if profile['publishMethod'] == 'MSDeploy':
+                scmUrl = profile['publishUrl'].replace(":443", "")
+                cd_url = 'https://' + profile['userName'] + ':' + profile['userPWD'] + '@' + scmUrl + '/docker/hook'
+                cd_settings['CI_CD_URL'] = cd_url
+                break
+    else:
+        cd_settings['CI_CD_URL'] = ''
+
+    settings.append(cd_settings)
+    return settings
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: I think you only need to return `cd_settings`

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py [hunk 13]
```diff
@@ -840,6 +840,44 @@ def list_publish_profiles(resource_group_name, name, slot=None):
     return converted
 
 
+def enable_cd(resource_group_name, name, enable, slot=None):
+    settings = []
+    settings.append("DOCKER_ENABLE_CI=" + enable)
+    settings_output = []
+    settings_output.append(('DOCKER_ENABLE_CI', enable))
```
Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: `settings_output ` is not used, and should be removed

File: src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py (hunk 12 above)
Comment (aspaan) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: A per your previous suggestion you mentioned to return this format
{
   "DOCKER_ENABLE_CI": "true or false",
   "CI_CD_URL" : "<url or empty>" 
}

should I not do that anymore?

Comment (yugangw-msft) [inline src/command_modules/azure-cli-appservice/azure/cli/command_modules/appservice/custom.py]: still do it, but just don't put it inside the array of `settings`  though. We are returning a dictionary, not an array, right?

**Comment by aspaan** [comment]
Hi Yugang, is there anything else I need on this?

**Comment by yugangw-msft** [comment]
> Hi Yugang, is there anything else I need on this?

Please address my review feedback and resolve conflicts

**Pull request merged**